File: backend/app/skills/builtin/acquisition/pubchem.py
"""PubChem acquisition skill — search and fetch compound data.

PubChem is the NIH database of chemical molecules and their activities.
This skill uses the three-tier fallback chain (api > httpx > crawl):

    1. API first: PUG-REST API (structured JSON/CSV)
    2. httpx second: direct page fetch with browser UA
    3. crawl fallback: Playwright-rendered visible page text

PUG-REST API docs: https://pubchem.ncbi.nlm.nih.gov/docs/pug-rest
"""

from __future__ import annotations

import asyncio
import json
import logging
from datetime import UTC, datetime
from typing import Any
from urllib.parse import quote

from agents import RunContextWrapper, function_tool
from bs4 import BeautifulSoup

from app.agent_loop.context import RunContext
from app.domain.contracts import (
    Database,
    DataLevel,
    QueryStatus,
    SourceRecord,
    generate_prefixed_uuid,
    make_source_id,
)
from app.skills.builtin.acquisition._download_io import download_file_for_run
from app.skills.registry import SkillCategory, SkillDef, skill_registry
from app.tools.crawler import CrawlAttempt, CrawlError, FetchResult, fetch_with_fallback

logger = logging.getLogger(__name__)

_PUGREST_BASE = "https://pubchem.ncbi.nlm.nih.gov/rest/pug"
_PUBCHEM_PAGE_BASE = "https://pubchem.ncbi.nlm.nih.gov/compound"
_MAX_BODY_CHARS = 5000


def _visible_text(html: str) -> str:
    soup = BeautifulSoup(html, "html.parser")
    for tag in soup(["script", "style", "head", "noscript"]):
        tag.decompose()
    return " ".join(soup.get_text(separator=" ", strip=True).split())


def _pubchem_properties(result: FetchResult) -> list[object] | None:
    if result.method_used == "api":
        try:
            data = json.loads(result.content)
        except (json.JSONDecodeError, TypeError):
            return None
        properties = data.get("PropertyTable", {}).get("Properties")
        return properties if isinstance(properties, list) else None
    return None


def _accept_pubchem_search_result(result: FetchResult) -> bool:
    if result.method_used == "api":
        return _pubchem_properties(result) is not None
    return result.method_used == "crawl" and bool(_visible_text(result.content))


def _accept_pubchem_compound_result(result: FetchResult) -> bool:
    if result.method_used == "api":
        return bool(_pubchem_properties(result))
    return result.method_used == "crawl" and bool(_visible_text(result.content))


def _attempt_audit(attempts: tuple[CrawlAttempt, ...]) -> list[dict[str, object]]:
    return [
        {
            "method": attempt.method,
            "url": attempt.url,
            "status": attempt.status,
            "status_code": attempt.status_code,
            "reason": attempt.reason,
            "fallback_reason": attempt.fallback_reason,
        }
        for attempt in attempts
    ]


def _page_fallback(source: str, page_url: str, result: FetchResult) -> str:
    return json.dumps(
        {
            "status": "page_fallback",
            "source": source,
            "method_used": result.method_used,
            "page_url": page_url,
            "body_text_preview": _visible_text(result.content)[:_MAX_BODY_CHARS],
            "attempts": _attempt_audit(result.attempts),
        },
        ensure_ascii=False,
    )


def _fallback_error(source: str, page_url: str, error: CrawlError) -> str:
    return json.dumps(
        {
            "status": "error",
            "source": source,
            "page_url": page_url,
            "attempted_methods": ["api", "httpx", "crawl"],
            "attempts": _attempt_audit(error.attempts),
            "error": str(error),
        },
        ensure_ascii=False,
    )


@function_tool(
    description_override=(
        "Search PubChem for chemical compounds matching a name or keyword. "
        "Parameters: ``term`` (required, search keyword like 'aspirin' or "
        "'caffeine'), ``max_results`` (optional, default 20). "
        "Returns JSON with compound records (CID, name, formula, MW, etc.). "
        "Use ``get_compound`` to get full details for a specific CID."
    ),
)
async def search_pubchem(
    ctx: RunContextWrapper[Any],
    term: str,
    max_results: int = 20,
) -> str:
    """Search PubChem for chemical compounds matching a name or keyword.

    Queries the PUG-REST API first. If the API is unavailable or cannot be
    parsed, rejects the static shell and uses Playwright-rendered page text.

    Args:
        ctx: Agent SDK run context wrapper.
        term: Compound name or keyword (e.g. "aspirin", "curcumin").
        max_results: Maximum number of compounds to return (default 20).

    Returns:
        JSON string with keys: source, term, count, records.
        On page fallback: JSON with status="page_fallback".
        On failure: JSON with status="error" and attempted methods.
    """
    run_ctx: RunContext = ctx.context
    encoded_term = quote(term)

    # PUG-REST: /compound/name/{name}/property/.../JSON
    api_url = (
        f"{_PUGREST_BASE}/compound/name/{encoded_term}/property/"
        f"MolecularFormula,MolecularWeight,IUPACName,CanonicalSMILES/"
        f"JSON?MaxRecords={max_results}"
    )

    page_url = f"https://pubchem.ncbi.nlm.nih.gov/#query={encoded_term}"
    try:
        result = await fetch_with_fallback(
            api_url,
            page_url,
            source_name="pubchem",
            accept_result=_accept_pubchem_search_result,
            facade=run_ctx.crawler_facade,
        )
    except CrawlError as exc:
        run_ctx.log_query(term, "pubchem", QueryStatus.FAILED, 0)
        return _fallback_error("pubchem", page_url, exc)
    if result.method_used == "api":
        try:
            data = json.loads(result.content)
            compounds = data.get("PropertyTable", {}).get("Properties", [])
            records = [
                {
                    "cid": c.get("CID", 0),
                    "molecular_formula": c.get("MolecularFormula", ""),
                    "molecular_weight": c.get("MolecularWeight", 0.0),
                    "iupac_name": c.get("IUPACName", ""),
                    "canonical_smiles": c.get("CanonicalSMILES", ""),
                    "url": f"{_PUBCHEM_PAGE_BASE}/{c.get('CID', '')}",
                }
                for c in compounds
            ]
            run_ctx.log_query(term, "pubchem", QueryStatus.SUCCESS, len(records))
            return json.dumps(
                {
                    "source": "pubchem",
                    "term": term,
                    "count": len(records),
                    "records": records,
                    "method_used": "api",
                    "attempts": _attempt_audit(result.attempts),
                },
                ensure_ascii=False,
            )
        except (json.JSONDecodeError, AttributeError, KeyError, TypeError) as exc:
            logger.warning("Failed to parse PubChem API response: %s", exc)
    run_ctx.log_query(term, "pubchem", QueryStatus.PAGE_FALLBACK, 0)
    return _page_fallback("pubchem", page_url, result)


@function_tool
async def get_compound(
    ctx: RunContextWrapper[Any],
    cid: int,
) -> str:
    """Get detailed information about a specific PubChem compound by CID.

    Queries the PUG-REST API for compound properties including molecular
    formula, weight, IUPAC name, SMILES, and InChI key.

    Args:
        ctx: Agent SDK run context wrapper.
        cid: PubChem Compound ID (e.g. 2244 for aspirin).

    Returns:
        JSON string with compound details.
        On page fallback: JSON with status="page_fallback".
        On failure: JSON with status="error" and attempted methods.
    """
    run_ctx: RunContext = ctx.context
    api_url = (
        f"{_PUGREST_BASE}/compound/cid/{cid}/property/"
        f"MolecularFormula,MolecularWeight,IUPACName,CanonicalSMILES,InChIKey,InChI/"
        f"JSON"
    )

    page_url = f"{_PUBCHEM_PAGE_BASE}/{cid}"
    try:
        result = await fetch_with_fallback(
            api_url,
            page_url,
            source_name="pubchem",
            accept_result=_accept_pubchem_compound_result,
            facade=run_ctx.crawler_facade,
        )
    except CrawlError as exc:
        run_ctx.log_query(str(cid), "pubchem", QueryStatus.FAILED, 0)
        return _fallback_error("pubchem", page_url, exc)
    if result.method_used == "api":
        try:
            data = json.loads(result.content)
            compounds = data.get("PropertyTable", {}).get("Properties", [])
            if compounds:
                c = compounds[0]
                record = {
                    "cid": c.get("CID", cid),
                    "molecular_formula": c.get("MolecularFormula", ""),
                    "molecular_weight": c.get("MolecularWeight", 0.0),
                    "iupac_name": c.get("IUPACName", ""),
                    "canonical_smiles": c.get("CanonicalSMILES", ""),
                    "inchi_key": c.get("InChIKey", ""),
                    "inchi": c.get("InChI", ""),
                    "url": f"{_PUBCHEM_PAGE_BASE}/{cid}",
                }
                run_ctx.log_query(str(cid), "pubchem", QueryStatus.SUCCESS, 1)

                source_record = SourceRecord(
                    source_id=make_source_id(Database.PUBCHEM, str(cid), api_url),
                    database=Database.PUBCHEM,
                    accession=str(cid),
                    url=api_url,
                    title=f"PubChem compound {cid}",
                    retrieved_at=datetime.now(UTC),
                )
                run_ctx.add_source(source_record)

                return json.dumps(
                    {
                        "source": "pubchem",
                        "cid": cid,
                        "record": record,
                        "method_used": "api",
                        "attempts": _attempt_audit(result.attempts),
                    },
                    ensure_ascii=False,
                )
        except (json.JSONDecodeError, AttributeError, KeyError, TypeError) as exc:
            logger.warning("Failed to parse PubChem compound response: %s", exc)
    run_ctx.log_query(str(cid), "pubchem", QueryStatus.PAGE_FALLBACK, 0)
    return _page_fallback("pubchem", page_url, result)


@function_tool
async def download_pubchem(
    ctx: RunContextWrapper[Any],
    cid: int,
    file_type: str = "sdf",
) -> str:
    """Download a PubChem compound structure file (SDF or MOL).

    Uses the PUG-REST ``record`` endpoint to fetch the full structure record
    in SDF (2D) or MOL (2D) format. The file is staged as a compliant
    ``SourceAsset`` in managed (subagent) runs and written to the task raw
    directory otherwise, then tracked via a ``SourceRecord`` for provenance.

    Args:
        ctx: Agent SDK run context wrapper.
        cid: PubChem Compound ID (e.g. 2244 for aspirin).
        file_type: ``sdf`` (default) or ``mol``.

    Returns:
        JSON with source, cid, source_url, local_files, format_hint, and
        retrieved_at (ISO-8601); ``error`` on failure.
    """
    run_ctx: RunContext = ctx.context
    file_type = file_type.lower().strip().lstrip(".")
    if file_type not in {"sdf", "mol"}:
        return json.dumps({
            "source": "pubchem",
            "cid": cid,
            "error": f"unsupported file_type: {file_type}. Use 'sdf' or 'mol'.",
        }, ensure_ascii=False)
    record_type = "2d"
    ext = file_type.upper()
    url = f"{_PUGREST_BASE}/compound/cid/{cid}/record/{ext}?record_type={record_type}"
    filename = f"CID{cid}.{file_type}"
    format_hint = f"pubchem_{file_type}"

    try:
        if run_ctx.subagent_id is not None:
            facade = run_ctx.crawler_facade_or_none
            if facade is None:
                raise RuntimeError("crawler facade is not bound to the child Run")
            result = await facade.download(url)
            if not result.ok:
                raise RuntimeError(result.error or f"HTTP {result.status_code}")
            content = result.content
            source_id = make_source_id(Database.PUBCHEM, str(cid), url)
            asset = await asyncio.to_thread(
                run_ctx.stage_source_asset,
                content=content,
                filename=filename,
                source_id=source_id,
                successful_attempt_id=generate_prefixed_uuid("download_attempt"),
                data_level=DataLevel.REPOSITORY_PROCESSED,
                media_type=(
                    "chemical/x-mdl-sdfile"
                    if file_type == "sdf"
                    else "chemical/x-mdl-molfile"
                ),
            )
            dest = run_ctx.source_asset_path(asset)
        else:
            dest = run_ctx.work_dir.raw_file(filename)
            await download_file_for_run(run_ctx, url, dest)
    except Exception as exc:
        return json.dumps({
            "source": "pubchem",
            "cid": cid,
            "source_url": url,
            "error": f"download failed: {exc}",
        }, ensure_ascii=False)

    local_files = [str(dest)]
    run_ctx.add_raw_asset(local_files[0])

    retrieved_at = datetime.now(UTC)
    source_record = SourceRecord(
        source_id=make_source_id(Database.PUBCHEM, str(cid), url),
        database=Database.PUBCHEM,
        accession=str(cid),
        url=url,
        title=f"PubChem compound {cid} {file_type.upper()} structure",
        retrieved_at=retrieved_at,
    )
    run_ctx.add_source(source_record)

    return json.dumps({
        "source": "pubchem",
        "cid": cid,
        "source_url": url,
        "local_files": local_files,
        "format_hint": format_hint,
        "retrieved_at": retrieved_at.isoformat(),
    }, ensure_ascii=False)


pubchem_skill = SkillDef(
    name="pubchem",
    category=SkillCategory.ACQUISITION,
    description=(
        "Search and fetch chemical compound data from PubChem. "
        "Use when the user asks about compounds, chemical structures, SMILES, "
        "molecular formulas, or needs compound properties by CID."
    ),
    instructions=(
        "Use search_pubchem to find compounds by name (e.g. 'aspirin', 'curcumin'). "
        "Use get_compound to fetch details for a specific compound by CID "
        "(e.g. 2244 for aspirin). Use download_pubchem to fetch the full SDF/MOL "
        "structure record for a CID. API failures automatically use a rendered "
        "page fallback and return a bounded visible-text preview."
    ),
    tools=[search_pubchem, get_compound, download_pubchem],
    supported_sources=["pubchem"],
    version="0.1.0",
)

skill_registry.register(pubchem_skill)

File: server/src/external/sources/pubchem.ts
/**
 * PubChem source client (Python ``skills/builtin/acquisition/pubchem.py``
 * parity): search and fetch compound data through PUG-REST with the
 * three-tier fallback chain (api → html → crawl).
 *
 * ``download_pubchem`` routes through ``acquireSource`` (P5-D3): the SDF/MOL
 * structure record is published as an immutable content-addressed SourceAsset
 * with a DownloadAttempt record.
 */

import path from "node:path";

import type { DataLevel } from "../../dataset/contracts/enums.js";
import type { SourceRecord } from "../../dataset/contracts/source.js";
import { acquireSource } from "../acquisition/downloader.js";
import type { ContentCache } from "../acquisition/content-cache.js";
import { isAbortError } from "../network/errors.js";
import type { PublicHttpClient } from "../network/http-client.js";
import type { SourceQueryContext } from "./context.js";
import {
  DEFAULT_RATE_LIMIT_MS,
  FallbackFailure,
  MAX_BODY_CHARS,
  fallbackFetch,
  isRecord,
  makeSourceId,
  quoteQuery,
  rateLimit,
  visibleText,
  type FallbackAttempt,
  type FallbackFetchResult,
} from "./fallback.js";

const PUGREST_BASE = "https://pubchem.ncbi.nlm.nih.gov/rest/pug";
const PUBCHEM_PAGE_BASE = "https://pubchem.ncbi.nlm.nih.gov/compound";

export interface DownloadDeps {
  taskRoot: string;
  cache: ContentCache;
  client: PublicHttpClient;
  signal?: AbortSignal;
  rateLimitMs?: number;
  maxDownloadBytes?: number;
  timeoutMs?: number;
}

/** Python ``_pubchem_properties``: the PropertyTable.Properties list or null. */
export function pubchemProperties(content: string): unknown[] | null {
  try {
    const data = JSON.parse(content) as unknown;
    if (!isRecord(data)) return null;
    const table = data["PropertyTable"];
    const properties = isRecord(table) ? table["Properties"] : undefined;
    return Array.isArray(properties) ? properties : null;
  } catch {
    return null;
  }
}

function attemptAudit(attempts: readonly FallbackAttempt[]): Array<Record<string, unknown>> {
  return attempts.map((attempt) => ({
    method: attempt.method,
    url: attempt.url,
    status: attempt.status,
    status_code: attempt.status_code,
    reason: attempt.reason,
    fallback_reason: attempt.fallback_reason,
  }));
}

function pageFallback(
  source: string,
  pageUrl: string,
  fetched: FallbackFetchResult,
): Record<string, unknown> {
  return {
    status: "page_fallback",
    source,
    method_used: fetched.method_used,
    page_url: pageUrl,
    body_text_preview: visibleText(fetched.content).slice(0, MAX_BODY_CHARS),
    attempts: attemptAudit(fetched.attempts),
  };
}

function fallbackError(
  source: string,
  pageUrl: string,
  error: FallbackFailure,
): Record<string, unknown> {
  return {
    status: "error",
    source,
    page_url: pageUrl,
    attempted_methods: ["api", "httpx", "crawl"],
    attempts: attemptAudit(error.attempts),
    error: error.message,
  };
}

function searchPubchemProperties(
  term: string,
  fetched: FallbackFetchResult,
  context: SourceQueryContext,
): Record<string, unknown> | null {
  try {
    const data = JSON.parse(fetched.content) as unknown;
    const document = isRecord(data) ? data : {};
    const table = isRecord(document["PropertyTable"]) ? document["PropertyTable"] : {};
    const compoundsRaw = Array.isArray(table["Properties"]) ? table["Properties"] : [];
    const records = compoundsRaw.map((raw) => {
      const compound = isRecord(raw) ? raw : {};
      const cidValue = compound["CID"];
      return {
        cid: typeof cidValue === "number" ? cidValue : 0,
        molecular_formula:
          typeof compound["MolecularFormula"] === "string" ? compound["MolecularFormula"] : "",
        molecular_weight:
          typeof compound["MolecularWeight"] === "number" ? compound["MolecularWeight"] : 0,
        iupac_name: typeof compound["IUPACName"] === "string" ? compound["IUPACName"] : "",
        canonical_smiles:
          typeof compound["CanonicalSMILES"] === "string" ? compound["CanonicalSMILES"] : "",
        url: `${PUBCHEM_PAGE_BASE}/${cidValue ?? ""}`,
      };
    });
    context.onQuery?.(term, "pubchem", "success", records.length);
    return {
      source: "pubchem",
      term,
      count: records.length,
      records,
      method_used: "api",
      attempts: attemptAudit(fetched.attempts),
    };
  } catch (error) {
    if (isAbortError(error) || context.signal?.aborted === true) throw error;
    // Python logs a warning and falls through to the page fallback with the api result.
    return null;
  }
}

/** Search PubChem for chemical compounds matching a name or keyword. */
export async function searchPubchem(
  term: string,
  maxResults: number,
  context: SourceQueryContext,
): Promise<Record<string, unknown>> {
  const encoded = quoteQuery(term);
  context.onQueryStarted?.(term, "pubchem");
  const apiUrl =
    `${PUGREST_BASE}/compound/name/${encoded}/property/` +
    `MolecularFormula,MolecularWeight,IUPACName,CanonicalSMILES/` +
    `JSON?MaxRecords=${maxResults}`;
  const pageUrl = `https://pubchem.ncbi.nlm.nih.gov/#query=${encoded}`;

  let fetched: Awaited<ReturnType<typeof fallbackFetch>>;
  try {
    fetched = await fallbackFetch({
      sourceName: "pubchem",
      apiUrl,
      pageUrl,
      // Python parity: the static html tier is always rejected for PubChem —
      // only the api tier and the rendered crawl tier can be accepted.
      acceptResult: (methodUsed, content) => {
        if (methodUsed === "api") return pubchemProperties(content) !== null;
        return methodUsed === "crawl" && visibleText(content).length > 0;
      },
      client: context.client,
      browserFallback: context.browserFallback,
      signal: context.signal,
      rateLimitMs: context.rateLimitMs,
    });
  } catch (error) {
    if (isAbortError(error) || context.signal?.aborted === true) throw error;
    if (!(error instanceof FallbackFailure)) throw error;
    context.onQuery?.(term, "pubchem", "failed", 0);
    return fallbackError("pubchem", pageUrl, error);
  }

  if (fetched.method_used === "api") {
    const parsed = searchPubchemProperties(term, fetched, context);
    if (parsed !== null) return parsed;
  }
  context.onQuery?.(term, "pubchem", "page_fallback", 0);
  return pageFallback("pubchem", pageUrl, fetched);
}

/** Get detailed information about a specific PubChem compound by CID. */
export async function getCompound(
  cid: number,
  context: SourceQueryContext,
): Promise<Record<string, unknown>> {
  context.onQueryStarted?.(String(cid), "pubchem");
  const apiUrl =
    `${PUGREST_BASE}/compound/cid/${cid}/property/` +
    `MolecularFormula,MolecularWeight,IUPACName,CanonicalSMILES,InChIKey,InChI/JSON`;
  const pageUrl = `${PUBCHEM_PAGE_BASE}/${cid}`;

  let fetched: Awaited<ReturnType<typeof fallbackFetch>>;
  try {
    fetched = await fallbackFetch({
      sourceName: "pubchem",
      apiUrl,
      pageUrl,
      // Python parity: an empty PropertyTable list is rejected for the
      // compound endpoint (bool(properties)), unlike the search endpoint.
      acceptResult: (methodUsed, content) => {
        if (methodUsed === "api") {
          const properties = pubchemProperties(content);
          return properties !== null && properties.length > 0;
        }
        return methodUsed === "crawl" && visibleText(content).length > 0;
      },
      client: context.client,
      browserFallback: context.browserFallback,
      signal: context.signal,
      rateLimitMs: context.rateLimitMs,
    });
  } catch (error) {
    if (isAbortError(error) || context.signal?.aborted === true) throw error;
    if (!(error instanceof FallbackFailure)) throw error;
    context.onQuery?.(String(cid), "pubchem", "failed", 0);
    return fallbackError("pubchem", pageUrl, error);
  }

  if (fetched.method_used === "api") {
    try {
      const properties = pubchemProperties(fetched.content);
      const first = properties !== null && properties.length > 0 ? properties[0] : undefined;
      if (isRecord(first)) {
        const record = {
          cid: typeof first["CID"] === "number" ? first["CID"] : cid,
          molecular_formula:
            typeof first["MolecularFormula"] === "string" ? first["MolecularFormula"] : "",
          molecular_weight: typeof first["MolecularWeight"] === "number" ? first["MolecularWeight"] : 0,
          iupac_name: typeof first["IUPACName"] === "string" ? first["IUPACName"] : "",
          canonical_smiles:
            typeof first["CanonicalSMILES"] === "string" ? first["CanonicalSMILES"] : "",
          inchi_key: typeof first["InChIKey"] === "string" ? first["InChIKey"] : "",
          inchi: typeof first["InChI"] === "string" ? first["InChI"] : "",
          url: `${PUBCHEM_PAGE_BASE}/${cid}`,
        };
        context.onQuery?.(String(cid), "pubchem", "success", 1);
        return {
          source: "pubchem",
          cid,
          record,
          method_used: "api",
          attempts: attemptAudit(fetched.attempts),
        };
      }
    } catch (error) {
      if (isAbortError(error) || context.signal?.aborted === true) throw error;
      // Python logs a warning and falls through to the page fallback.
    }
  }
  context.onQuery?.(String(cid), "pubchem", "page_fallback", 0);
  return pageFallback("pubchem", pageUrl, fetched);
}

/** Download a PubChem compound structure file (SDF or MOL) via acquireSource. */
export async function downloadPubchem(
  cid: number,
  fileType: string,
  deps: DownloadDeps,
): Promise<Record<string, unknown>> {
  const normalizedType = fileType.toLowerCase().trim().replace(/^\./, "");
  if (normalizedType !== "sdf" && normalizedType !== "mol") {
    return {
      source: "pubchem",
      cid,
      error: `unsupported file_type: ${normalizedType}. Use 'sdf' or 'mol'.`,
    };
  }
  const ext = normalizedType.toUpperCase();
  const url = `${PUGREST_BASE}/compound/cid/${cid}/record/${ext}?record_type=2d`;
  const filename = `CID${cid}.${normalizedType}`;
  const formatHint = `pubchem_${normalizedType}`;

  const retrievedAt = new Date().toISOString();
  const source: SourceRecord = {
    schema_version: "1.0",
    source_id: makeSourceId("pubchem", String(cid), url),
    database: "pubchem",
    accession: String(cid),
    url,
    title: `PubChem compound ${cid} ${ext} structure`,
    retrieved_at: retrievedAt,
  };

  try {
    await rateLimit(deps.rateLimitMs ?? DEFAULT_RATE_LIMIT_MS);
    const result = await acquireSource({
      source,
      filename,
      workdirRoot: deps.taskRoot,
      cache: deps.cache,
      client: deps.client,
      dataLevel: "repository_processed" satisfies DataLevel,
      maxBytes: deps.maxDownloadBytes ?? 4096 * 1024 * 1024,
      accept: "*/*",
      signal: deps.signal,
      timeoutMs: deps.timeoutMs,
    });
    if (result.asset === null) {
      return {
        source: "pubchem",
        cid,
        source_url: url,
        error: `download failed: ${result.attempt.error_message ?? "unknown error"}`,
      };
    }
    return {
      source: "pubchem",
      cid,
      source_url: url,
      local_files: [path.join(deps.taskRoot, result.asset.relative_path)],
      format_hint: formatHint,
      retrieved_at: retrievedAt,
    };
  } catch (error) {
    if (isAbortError(error) || deps.signal?.aborted === true) throw error;
    return {
      source: "pubchem",
      cid,
      source_url: url,
      error: `download failed: ${error instanceof Error ? error.message : String(error)}`,
    };
  }
}
